In [38]:
import pandas as pd
import os

DATA_DIR = "../data/reduced"

customers = pd.read_csv(
    os.path.join(DATA_DIR, "customers_reduced.csv")
)

articles = pd.read_csv(
    os.path.join(DATA_DIR, "articles_reduced.csv")
)

transactions = pd.read_csv(
    os.path.join(DATA_DIR, "transactions_reduced.csv")
)

print("Customers:", customers.shape)
print("Articles:", articles.shape)
print("Transactions:", transactions.shape)

import os

# Create processed folder if it doesn't exist
os.makedirs("../data/processed", exist_ok=True)

# Save cleaned transaction + article dataset
output_path = "../data/processed/transactions_articles_clean.csv"

transaction_articles.to_csv(output_path, index=False)

print(f"✅ Cleaned dataset saved to: {output_path}")
print(f"Dataset shape: {transaction_articles.shape}")

Customers: (51527, 7)
Articles: (19519, 25)
Transactions: (200000, 5)
✅ Cleaned dataset saved to: ../data/processed/transactions_articles_clean.csv
Dataset shape: (200000, 29)


In [39]:
customers.head()

,customer_id,FN,Active,club_member_status,fashion_news_frequency,age,postal_code
0,0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...,NaN,NaN,ACTIVE,NONE,25.0,2973abc54daa8a5f8ccfe9362140c63247c5eee03f1d93...
1,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,NaN,NaN,ACTIVE,NONE,24.0,64f17e6a330a85798e4998f62d0930d14db8db1c054af6...
2,00007d2de826758b65a93dd24ce629ed66842531df6699...,1.0,1.0,ACTIVE,Regularly,32.0,8d6f45050876d059c830a0fe63f1a4c022de279bb68ce3...
3,00025f8226be50dcab09402a2cacd520a99e112fe01fdd...,1.0,1.0,ACTIVE,Regularly,25.0,a04505cd362b8b1e3f1d8d47d8b31163eae82f078182a1...
4,0003abe64294e66a6310c3436fa9e5b754cc5603deef4f...,NaN,NaN,ACTIVE,NONE,51.0,06ef7695ef26ebd1fe08304240e9e32714a1ee85c6c348...


In [40]:
articles.head()

,article_id,product_code,prod_name,product_type_no,product_type_name,product_group_name,graphical_appearance_no,graphical_appearance_name,colour_group_code,colour_group_name,...,department_name,index_code,index_name,index_group_no,index_group_name,section_no,section_name,garment_group_no,garment_group_name,detail_desc
0,108775015,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,9,Black,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
1,108775044,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,10,White,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
2,108775051,108775,Strap top (1),253,Vest top,Garment Upper body,1010017,Stripe,11,Off White,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
3,110065001,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,9,Black,...,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulde..."
4,110065002,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,10,White,...,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulde..."


In [41]:
transactions.head()

,t_dat,customer_id,article_id,price,sales_channel_id
0,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,663713001,0.050831,2
1,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,541518023,0.030492,2
2,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,505221004,0.015237,2
3,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687003,0.016932,2
4,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687004,0.016932,2


In [42]:
customers.info()

<class 'pandas.DataFrame'>
RangeIndex: 51527 entries, 0 to 51526
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   customer_id             51527 non-null  str    
 1   FN                      21423 non-null  float64
 2   Active                  21011 non-null  float64
 3   club_member_status      50844 non-null  str    
 4   fashion_news_frequency  51281 non-null  str    
 5   age                     51230 non-null  float64
 6   postal_code             51527 non-null  str    
dtypes: float64(3), str(4)
memory usage: 2.8 MB


In [43]:
customers.describe(include="all")

,customer_id,FN,Active,club_member_status,fashion_news_frequency,age,postal_code
count,51527,21423.0,21011.0,50844,51281,51230.000000,51527
unique,51527,NaN,NaN,3,3,NaN,45703
top,0000423b00ade91418cceaf3b26c6af3dd342b51fd051e...,NaN,NaN,ACTIVE,NONE,NaN,2c29ae653a9282cce4151bd87643c907644e09541abc28...
freq,1,NaN,NaN,49464,29787,NaN,295
mean,NaN,1.0,1.0,NaN,NaN,37.397989,NaN
std,NaN,0.0,0.0,NaN,NaN,13.573193,NaN
min,NaN,1.0,1.0,NaN,NaN,17.000000,NaN
25%,NaN,1.0,1.0,NaN,NaN,26.000000,NaN
50%,NaN,1.0,1.0,NaN,NaN,33.000000,NaN
75%,NaN,1.0,1.0,NaN,NaN,49.000000,NaN


In [44]:
customers.columns.tolist()

['customer_id',
 'FN',
 'Active',
 'club_member_status',
 'fashion_news_frequency',
 'age',
 'postal_code']

In [45]:
print("Number of customers:", customers["customer_id"].nunique())

Number of customers: 51527


In [46]:
articles.info()

<class 'pandas.DataFrame'>
RangeIndex: 19519 entries, 0 to 19518
Data columns (total 25 columns):
 #   Column                        Non-Null Count  Dtype
---  ------                        --------------  -----
 0   article_id                    19519 non-null  int64
 1   product_code                  19519 non-null  int64
 2   prod_name                     19519 non-null  str  
 3   product_type_no               19519 non-null  int64
 4   product_type_name             19519 non-null  str  
 5   product_group_name            19519 non-null  str  
 6   graphical_appearance_no       19519 non-null  int64
 7   graphical_appearance_name     19519 non-null  str  
 8   colour_group_code             19519 non-null  int64
 9   colour_group_name             19519 non-null  str  
 10  perceived_colour_value_id     19519 non-null  int64
 11  perceived_colour_value_name   19519 non-null  str  
 12  perceived_colour_master_id    19519 non-null  int64
 13  perceived_colour_master_name  19519 non-nu

In [47]:
articles.columns.tolist()

['article_id',
 'product_code',
 'prod_name',
 'product_type_no',
 'product_type_name',
 'product_group_name',
 'graphical_appearance_no',
 'graphical_appearance_name',
 'colour_group_code',
 'colour_group_name',
 'perceived_colour_value_id',
 'perceived_colour_value_name',
 'perceived_colour_master_id',
 'perceived_colour_master_name',
 'department_no',
 'department_name',
 'index_code',
 'index_name',
 'index_group_no',
 'index_group_name',
 'section_no',
 'section_name',
 'garment_group_no',
 'garment_group_name',
 'detail_desc']

In [48]:
articles.head()

,article_id,product_code,prod_name,product_type_no,product_type_name,product_group_name,graphical_appearance_no,graphical_appearance_name,colour_group_code,colour_group_name,...,department_name,index_code,index_name,index_group_no,index_group_name,section_no,section_name,garment_group_no,garment_group_name,detail_desc
0,108775015,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,9,Black,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
1,108775044,108775,Strap top,253,Vest top,Garment Upper body,1010016,Solid,10,White,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
2,108775051,108775,Strap top (1),253,Vest top,Garment Upper body,1010017,Stripe,11,Off White,...,Jersey Basic,A,Ladieswear,1,Ladieswear,16,Womens Everyday Basics,1002,Jersey Basic,Jersey top with narrow shoulder straps.
3,110065001,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,9,Black,...,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulde..."
4,110065002,110065,OP T-shirt (Idro),306,Bra,Underwear,1010016,Solid,10,White,...,Clean Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Microfibre T-shirt bra with underwired, moulde..."


In [49]:
print("Number of articles:", articles["article_id"].nunique())

Number of articles: 19519


In [50]:
articles.nunique().sort_values(ascending=False)

article_id                      19519
prod_name                       10813
product_code                    10584
detail_desc                      9900
department_no                     274
department_name                   228
product_type_no                    96
product_type_name                  95
section_no                         56
section_name                       56
colour_group_name                  50
colour_group_code                  50
graphical_appearance_no            29
graphical_appearance_name          29
garment_group_no                   21
garment_group_name                 21
perceived_colour_master_id         19
perceived_colour_master_name       19
product_group_name                 14
index_code                         10
index_name                         10
perceived_colour_value_id           8
perceived_colour_value_name         8
index_group_name                    5
index_group_no                      5
dtype: int64

In [51]:
transactions.info()

<class 'pandas.DataFrame'>
RangeIndex: 200000 entries, 0 to 199999
Data columns (total 5 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   t_dat             200000 non-null  str    
 1   customer_id       200000 non-null  str    
 2   article_id        200000 non-null  int64  
 3   price             200000 non-null  float64
 4   sales_channel_id  200000 non-null  int64  
dtypes: float64(1), int64(2), str(2)
memory usage: 7.6 MB


In [52]:
transactions.head()

,t_dat,customer_id,article_id,price,sales_channel_id
0,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,663713001,0.050831,2
1,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,541518023,0.030492,2
2,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,505221004,0.015237,2
3,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687003,0.016932,2
4,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687004,0.016932,2


In [53]:
transactions.columns.tolist()

['t_dat', 'customer_id', 'article_id', 'price', 'sales_channel_id']

In [54]:
print("Unique customers:", transactions["customer_id"].nunique())
print("Unique articles:", transactions["article_id"].nunique())

Unique customers: 51527
Unique articles: 19519


In [55]:
transactions["t_dat"] = pd.to_datetime(transactions["t_dat"])

print("Minimum date:", transactions["t_dat"].min())
print("Maximum date:", transactions["t_dat"].max())

Minimum date: 2018-09-20 00:00:00
Maximum date: 2018-09-24 00:00:00


In [56]:
print("Duplicate customer rows:", customers.duplicated().sum())
print("Duplicate article rows:", articles.duplicated().sum())
print("Duplicate transaction rows:", transactions.duplicated().sum())

Duplicate customer rows: 0
Duplicate article rows: 0
Duplicate transaction rows: 19741


In [57]:
transactions["t_dat"].value_counts().sort_index().head(20)

t_dat
2018-09-20    48399
2018-09-21    47543
2018-09-22    17635
2018-09-23    52230
2018-09-24    34193
Name: count, dtype: int64

In [58]:
top_articles = (
    transactions["article_id"]
    .value_counts()
    .head(10)
)

top_articles

article_id
685687001    1109
685687003    1093
685687004    1093
685687002     758
562245001     569
539723005     411
562245018     403
573085001     392
573085020     386
524825013     378
Name: count, dtype: int64

In [59]:
top_customers = (
    transactions["customer_id"]
    .value_counts()
    .head(10)
)

top_customers

customer_id
2fdf822dbaad2b983b37e651a982bba24352a92c8a5c4c75be25c771f2af6d13    103
a81e0b7657a090198d8138c95fae7db1218305b97bb8fd3f42b8d8e1b6d5b846    100
4aa23f5f3013303a7f194972fd68cf5cb04cdc88ab69bee8acf06283e7988e15     80
f6fbb1480291047c308f83a25e8e8b1e2c5a44a9390662b1d8435f700b5555db     72
3657f79cf5f3622e4506a3a9c91bb610572ad2d34458b12abdd7c06274d1980b     55
55e7d8de6acc36f8e77d681ccfab65dd67bae06cd4e2889b44f472f033aca0e8     54
05c5685f1301de31ac4463293228f74ba1b3b427dc1a79dbf54f0cf98f242e81     52
31287b3d29b025cf00822b66b462a415e9c58d6538562709292c54d2bb6af25a     52
4043eeb9fb65b1735639ce44efbc5c43d2f9160fb26c8da6d73ed2e211a01fdf     52
4fa96051d6715de656891c600846990aa7e224086969f0c690eec96b4b3fe6eb     51
Name: count, dtype: int64

In [60]:
transaction_articles = transactions.merge(
    articles,
    on="article_id",
    how="left"
)

transaction_articles.head()

,t_dat,customer_id,article_id,price,sales_channel_id,product_code,prod_name,product_type_no,product_type_name,product_group_name,...,department_name,index_code,index_name,index_group_no,index_group_name,section_no,section_name,garment_group_no,garment_group_name,detail_desc
0,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,663713001,0.050831,2,663713,Atlanta Push Body Harlow,283,Underwear body,Underwear,...,Expressive Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Lace push-up body with underwired, moulded, pa..."
1,2018-09-20,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,541518023,0.030492,2,541518,Rae Push (Melbourne) 2p,306,Bra,Underwear,...,Casual Lingerie,B,Lingeries/Tights,1,Ladieswear,61,Womens Lingerie,1017,"Under-, Nightwear","Lace push-up bras with underwired, moulded, pa..."
2,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,505221004,0.015237,2,505221,Inca Jumper,252,Sweater,Garment Upper body,...,Tops Knitwear DS,D,Divided,2,Divided,58,Divided Selected,1003,Knitwear,Jumper in rib-knit cotton with hard-worn detai...
3,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687003,0.016932,2,685687,W YODA KNIT OL OFFER,252,Sweater,Garment Upper body,...,Campaigns,A,Ladieswear,1,Ladieswear,15,Womens Everyday Collection,1023,Special Offers,V-neck knitted jumper with long sleeves and ri...
4,2018-09-20,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687004,0.016932,2,685687,W YODA KNIT OL OFFER,252,Sweater,Garment Upper body,...,Campaigns,A,Ladieswear,1,Ladieswear,15,Womens Everyday Collection,1023,Special Offers,V-neck knitted jumper with long sleeves and ri...


In [61]:
print(transaction_articles.shape)

(200000, 29)


In [62]:
print("Missing article IDs:", transaction_articles["product_code"].isna().sum())

Missing article IDs: 0


In [63]:
transaction_articles[
    [
        "customer_id",
        "article_id",
        "price",
        "t_dat",
        "product_code",
        "product_type_no",
        "product_group_name",
        "index_name",
        "index_group_name",
        "colour_group_name"
    ]
].head()

,customer_id,article_id,price,t_dat,product_code,product_type_no,product_group_name,index_name,index_group_name,colour_group_name
0,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,663713001,0.050831,2018-09-20,663713,283,Underwear,Lingeries/Tights,Ladieswear,Black
1,000058a12d5b43e67d225668fa1f8d618c13dc232df0ca...,541518023,0.030492,2018-09-20,541518,306,Underwear,Lingeries/Tights,Ladieswear,Light Pink
2,00007d2de826758b65a93dd24ce629ed66842531df6699...,505221004,0.015237,2018-09-20,505221,252,Garment Upper body,Divided,Divided,Pink
3,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687003,0.016932,2018-09-20,685687,252,Garment Upper body,Ladieswear,Ladieswear,Pink
4,00007d2de826758b65a93dd24ce629ed66842531df6699...,685687004,0.016932,2018-09-20,685687,252,Garment Upper body,Ladieswear,Ladieswear,Dark Green


In [64]:
transaction_articles.isnull().sum().sort_values(ascending=False).head(15)

detail_desc                  2141
customer_id                     0
article_id                      0
price                           0
t_dat                           0
product_code                    0
prod_name                       0
product_type_no                 0
product_type_name               0
product_group_name              0
graphical_appearance_no         0
graphical_appearance_name       0
sales_channel_id                0
colour_group_code               0
colour_group_name               0
dtype: int64

In [65]:
missing_percentage = (
    transaction_articles.isnull().sum()
    / len(transaction_articles)
    * 100
)

missing_percentage[missing_percentage > 0].sort_values(ascending=False)

detail_desc    1.0705
dtype: float64

In [66]:
transaction_articles["detail_desc"] = (
    transaction_articles["detail_desc"]
    .fillna("No description available")
)

In [67]:
print(
    "Missing detail_desc after handling:",
    transaction_articles["detail_desc"].isna().sum()
)

Missing detail_desc after handling: 0


In [69]:
import sys
import os

sys.path.append(os.path.abspath(".."))

from src.data_processing.missing_values import handle_missing_values

transaction_articles = handle_missing_values(transaction_articles)

Missing values before handling:
Series([], dtype: int64)

Missing values after handling:
Series([], dtype: int64)


In [70]:
import os

os.makedirs("../data/processed", exist_ok=True)

output_path = "../data/processed/transactions_articles_clean.csv"

transaction_articles.to_csv(output_path, index=False)

print("✅ Processed dataset saved successfully!")
print("File:", output_path)
print("Shape:", transaction_articles.shape)

✅ Processed dataset saved successfully!
File: ../data/processed/transactions_articles_clean.csv
Shape: (200000, 29)


In [71]:
import pandas as pd

clean_data = pd.read_csv(
    "../data/processed/transactions_articles_clean.csv"
)

print("Shape:", clean_data.shape)
print("Missing values:", clean_data.isnull().sum().sum())

Shape: (200000, 29)
Missing values: 0


# Overall Data Exploration Summary

## 1. Dataset Overview

The reduced H&M dataset was used for exploratory data analysis. The reduced transaction dataset contains **200,000 transactions**, along with corresponding customer and article information.

After merging the transaction data with the article data using `article_id`, the resulting dataset contains:

- **Rows:** 200,000
- **Columns:** 29

## 2. Transaction Date Analysis

The transaction data was examined based on transaction dates (`t_dat`).

The earliest transactions in the reduced dataset begin from **2018-09-20**, with the transaction volume varying across different dates.

The date distribution was explored using transaction counts grouped by `t_dat`.

## 3. Most Frequently Purchased Articles

The most frequently purchased articles were identified using `article_id` value counts.

The analysis showed that some articles occur significantly more frequently than others, indicating differences in article popularity within the reduced transaction dataset.

## 4. Most Active Customers

Customer purchase frequency was analyzed using `customer_id` value counts.

The analysis identified customers with the highest number of transactions, showing that customer activity is not uniformly distributed across the dataset.

## 5. Transaction and Article Data Integration

The transaction dataset was merged with the article dataset using:

- **Key:** `article_id`
- **Join type:** Left join

The resulting dataset contains transaction-level information together with article attributes such as:

- `product_code`
- `product_type_no`
- `product_type_name`
- `product_group_name`
- `graphical_appearance_no`
- `graphical_appearance_name`
- `colour_group_code`
- `colour_group_name`
- `index_name`
- `index_group_name`
- `department_no`
- `department_name`
- `section_no`
- `section_name`
- `garment_group_no`
- `garment_group_name`
- `detail_desc`

The merged dataset contains **200,000 rows and 29 columns**.

## 6. Missing Value Analysis

Missing values were checked after merging the datasets.

The analysis initially identified missing values in the `detail_desc` column:

- **Missing values:** 2,141
- **Missing percentage:** 1.0705%

No missing article IDs were found in the merged dataset.

## 7. Missing Value Handling

The missing values in `detail_desc` were handled by replacing them with:

`"No description available"`

After handling the missing values, the dataset was verified again.

**Result:**
- Missing `detail_desc` values after handling: **0**
- Total missing values in the processed dataset: **0**

## 8. Processed Dataset

The cleaned transaction-article dataset was saved as:

`data/processed/transactions_articles_clean.csv`

Final dataset dimensions:

- **200,000 rows**
- **29 columns**
- **0 missing values**

## Conclusion

The exploratory analysis provided an overview of transaction dates, article popularity, customer activity, and the relationship between transaction and article data. The transaction and article datasets were successfully integrated, missing article descriptions were handled, and the resulting cleaned dataset was saved for use in subsequent recommendation-engine processing and modeling tasks.